### PR TITLE
Refactor component loaders

### PR DESCRIPTION
**Pull request by bergie**
The goal is to remove dynamic requires from `noflo.ComponentLoader`, so NoFlo works better with module loaders like webpack.

Related to webpack/webpack#1299

### CODE REVIEW

**Comment by bergie** [comment]
@jonnor ready for review

File: spec/Subgraph.coffee [hunk 1]
```diff
@@ -209,6 +209,7 @@ describe 'Graph component', ->
     gr.addOutport 'out', 'Split', 'out'
     it 'should emit a ready event after network has been loaded', (done) ->
       @timeout 6000
+      c.baseDir = root
```
Comment (jonnor) [inline spec/Subgraph.coffee]: Why is this required now?

Comment (bergie) [inline spec/Subgraph.coffee]: Because some additional error handling was added to the Subgraph component. Previously these had failed, but silently

File: src/components/Graph.coffee [hunk 2]
```diff
@@ -24,13 +24,15 @@ class Graph extends noflo.Component
     if typeof graph is 'object'
       if typeof graph.addNode is 'function'
         # Existing Graph object
-        return @createNetwork graph
+        return @createNetwork graph, (err) =>
+          return @error err if err
```
Comment (jonnor) [inline src/components/Graph.coffee]: Is this and the below error handling needed here, or just added since passed over this area of code?

Comment (bergie) [inline src/components/Graph.coffee]: Just added because I happened to look at it and noticed errors were ignoredd

File: src/lib/ComponentLoader.coffee [hunk 3]
```diff
@@ -74,16 +32,16 @@ class ComponentLoader extends EventEmitter
...
-    setTimeout =>
-      @components = {}
-
-      @getModuleComponents @baseDir
 
+    @components = {}
+    registerLoader.register @, (err) =>
+      if err
+        return callback err if err
+        throw err
```
Comment (bergie) [inline src/lib/ComponentLoader.coffee]: Correct, I'll fix that

**Comment by jonnor** [comment]
This code allows to use NoFlo in Webpack builds for Node.js (browser not tested).
Here are the basic snippets of the webpack config needed:

```
  module: {
    noParse: [
      /loader\/NodeJs/,
      /loader\/ComponentIo/
    ]
  }
```

```
  plugins: [
    new webpack.NormalModuleReplacementPlugin(/loader\/register/,
                                              require.resolve('./staticloader'))
  ]
```

```
  externals: {
    'noflo': 'commonjs noflo'
  }
```

File: src/lib/loader/NodeJs.coffee [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+  return callback null unless componentLoaders.length
+  done = _.after componentLoaders.length, callback
+  componentLoaders.forEach (loaderPath) =>
+    cLoader = require loaderPath
+    loader.registerLoader cLoader, (err) ->
+      return callback err if err
+      done null
+
+manifestLoader =
+  writeCache: (loader, options, manifest, callback) ->
```
Comment (jonnor) [inline src/lib/loader/NodeJs.coffee]: I don't think we should do this implicitly on `readCache()`. Only `noflo-cache-preheat` should write it

Comment (bergie) [inline src/lib/loader/NodeJs.coffee]: That is something we could indeed do quite easily

Comment (bergie) [inline src/lib/loader/NodeJs.coffee]: And of course the nice thing is that the manifest file can just as well be generated using `fbp-manifest-deps` or other similar tools
